wire-grpc-client/src/main/java/com/squareup/wire/GrpcMethod.kt

```diff
@@ -0,0 +1,276 @@
+    val requestChannel = Channel<S>(1)
+    val responseChannel = Channel<R>(1)
+
+    val requestBody = newDuplexRequestBody()
+    val call = grpcClient.newCall(path, requestBody)
+    requestChannel.writeToRequestBody(requestBody, requestAdapter)
+    call.enqueue(responseChannel.readFromResponseBodyCallback(responseAdapter))
+
+    responseChannel.invokeOnClose { cause ->
+      call.cancel()
+      requestChannel.cancel()
+      responseChannel.cancel()
+    }
+
+    return requestChannel to responseChannel
+  }
+
+  private fun callBlocking(
+    grpcClient: GrpcClient,
+    requestBody: RequestBody
+  ): MessageSource<R> {
+    val call = grpcClient.newCall(path, requestBody)
+    val messageSource = BlockingMessageSource(call, responseAdapter)
+    call.enqueue(messageSource.readFromResponseBodyCallback())
+    return messageSource
+  }
+
+  /** Single request, single response. */
+  class RequestResponse<S : Any, R : Any>(
+    path: String,
+    requestAdapter: ProtoAdapter<S>,
+    responseAdapter: ProtoAdapter<R>
+  ) : GrpcMethod<S, R>(path, requestAdapter, responseAdapter) {
+    override fun invoke(grpcClient: GrpcClient, args: Array<Any>): Any {
+      return (args.last() as Continuation<Any>).invokeSuspending {
+        invoke(grpcClient, parameter = args[0])
+      }
+    }
+
+    suspend fun invoke(
+      grpcClient: GrpcClient,
+      parameter: Any
+    ): R {
+      val (requestChannel, responseChannel) = super.callWithChannels(grpcClient)
+
+      responseChannel.consume {
+        requestChannel.send(parameter as S)
+        requestChannel.close()
+        return receive()
+      }
+    }
+  }
+
+  /** Single request, single response. */
+  class BlockingRequestResponse<S : Any, R : Any>(
+    path: String,
+    requestAdapter: ProtoAdapter<S>,
+    responseAdapter: ProtoAdapter<R>
+  ) : GrpcMethod<S, R>(path, requestAdapter, responseAdapter) {
+    override fun invoke(grpcClient: GrpcClient, args: Array<Any>): Any {
+      val requestBody = newRequestBody(requestAdapter, args[0] as S)
+      val messageSource = super.callBlocking(grpcClient, requestBody)
+
+      messageSource.use {
+        return messageSource.read() ?: throw ProtocolException("required message not sent")
+      }
+    }
+  }
+
+  /** Request is streaming, with one single response. */
+  class StreamingRequest<S : Any, R : Any>(
+    path: String,
+    requestAdapter: ProtoAdapter<S>,
+    responseAdapter: ProtoAdapter<R>
+  ) : GrpcMethod<S, R>(path, requestAdapter, responseAdapter) {
+    override fun invoke(
+      grpcClient: GrpcClient,
+      args: Array<Any>
+    ): Pair<SendChannel<S>, Deferred<R>> {
+      val (requestChannel, responseChannel) = super.callWithChannels(grpcClient)
+
+      return Pair(
+          requestChannel,
+          // We use ATOMIC to guarantee the coroutine doesn't start after potential cancellations
+          // happen.
+          unconfinedCoroutineScope.async(start = CoroutineStart.ATOMIC) {
```

Comment:
Need to wrap my head around this more, and this is also copied from existing code, but this feels like another case of scope misuse.

Reply:
Yeah, we were learning as we went on this.